content/post/080-big-refactors.md

```diff
@@ -0,0 +1,26 @@
+---
+date: 2020-01-28
+url: 2020-01-28-big-refactors
+title: Big Refactors
+author: Alan Shaw
+---
+
+> If you're considering a refactor that'll touch **~70** interdependent repos you've come to the right place for some perspective! This is the sister post to [The Async Await Refactor](/2020-01-28-async-await-refactor) and covers some of the learnings of completing a big code refactor with a distributed team.
+
+If you think you'd like to take on a challenge like this, do the math. If each repo takes a day to refactor and if you work on it 5 days a week then you've automatically used 14 of your 52 weeks in a year. That's like, 4 months, and that's the absolute best case scenario.
+
+In reality, even if you get approval to do it you're unlikely to be able to work on it all day every day. The problem is that the benefits of a refactor can't always be seen until everything is done and the whole stack is using the new code. The point is, **your project needs to keep moving**. New features need to land and those bugs don't squash themselves.
+
+Refactors to each dependent repo **won't take just one day**. A tiny percentage of them will take less than a day but the vast majority will take multiple days to complete, if not weeks.
+
+You'll be tempted to make those API changes and performance refactors you've always wanted to make. **Beware of scope creep**. If you also decide to take on these tasks, be aware that you're adding to the complexity of integrating a massive breaking change and also to the time the refactor is going to take. You'll probably significantly underestimate the amount of extra time your additional changes will take to write, be tested, be reviewed and re-integrated into the project. **Stay focused**. You cannot make everything perfect now. Pick your battles, open an issue, and move on.
```

Comment:
You are also likely to introduce new bugs with your 'improvements' and not spend the same sort of time thinking about them because you are distracted by your flow control refactor.